rosbag2/include/rosbag2/reader.hpp

```diff
@@ -0,0 +1,162 @@
+    std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory =
+    std::make_shared<SerializationFormatConverterFactory>());
+
+  virtual ~Reader();
+
+  /**
+   * Close the storage object. Calls the reset function of the storage implementation which
+   * should close database connections.
+   */
+  void reset();
+
+  /**
+   * Open a rosbag for reading messages sequentially (time-ordered). Throws if file could not be
+   * opened. This must be called before any other function is used. The rosbag is
+   * automatically closed on destruction.
+   *
+   * If the `output_serialization_format` within the `converter_options` is not the same as the
+   * format of the underlying stored data, a converter will be used to automatically convert the
+   * data to the specified output format.
+   * Throws if the converter plugin does not exist.
+   *
+   * \param storage_options Options to configure the storage
+   * \param converter_options Options for specifying the output data format
+   */
+  virtual void open(
+    const StorageOptions & storage_options, const ConverterOptions & converter_options);
+
+  /**
+   * Ask whether the underlying bagfile contains at least one more message.
+   *
+   * \return true if storage contains at least one more message
+   * \throws runtime_error if the Reader is not open.
+   */
+  virtual bool has_next() const;
+
+  /**
+   * Read next message from storage. Will throw if no more messages are available.
+   * The message will be serialized in the format given to `open`.
+   *
+   * Expected usage:
+   * if (reader.has_next()) message = reader.read_next();
+   *
+   * \return next message in serialized form
+   * \throws runtime_error if the Reader is not open.
+   */
+  virtual std::shared_ptr<SerializedBagMessage> read_next();
+
+  /**
+   * Ask bagfile for all topics (including their type identifier) that were recorded.
+   *
+   * \return vector of topics with topic name and type as std::string
+   * \throws runtime_error if the Reader is not open.
+   */
+  virtual std::vector<TopicMetadata> get_all_topics_and_types() const;
+
+  /**
+   * Open a new database file to read from.
+   *
+   * \return true if there is a file to read from. false otherwise.
+   */
+  bool open_read_only(const std::string & file, const std::string & storage_id);
+
+  /**
+   * Gets a copy of the metadata file read from the storage factory.
+   *
+   * \return BagMetadata object with the info read from a metadata.yaml file
+   */
+  // TODO(piraka9011): Do we want to optimize to avoid copying?
+  rosbag2_storage::BagMetadata get_metadata() const;
+
+private:
+  /**
+   * Checks if all topics in the bagfile have the same RMW serialization format.
+   * Currently a bag file can only be played if all topics have the same serialization format.
+   *
+   * \param topics Vector of TopicInformation with metadata.
+   * \throws runtime_error if any topic has a different serialization format from the rest.
```

Comment:
I am not sure if I agree with that. It should definitely be possible for rosbags to store topics where the serialization format is not the same. So for example it's a valid use case to enhance a bag file with raw UDP captures from a sensor (i.e. velodyne) and store it with a separate serialization format. 
The requirement can be though that the serialization format doesn't change per topic.